Duckling/Recurrence/EN/Rules.hs

```diff
@@ -15,133 +15,247 @@ module Duckling.Recurrence.EN.Rules
-        Just . Token Recurrence $ recurrence grain 2
+      (Token RegexMatch (GroupMatch (lead:grain:_)):
+       _) -> do
+          (v, t) <- case Text.toLower lead of
+            "bi"   -> Just (2, 1)
+            "semi" -> Just (1, 2)
+            _      -> Just (1, 1)
+          g <- case Text.toLower grain of
+            "hour"    -> Just TG.Hour
+            "dai"     -> Just TG.Day
+            "week"    -> Just TG.Week
+            "month"   -> Just TG.Month
+            "quarter" -> Just TG.Quarter
+            "year"    -> Just TG.Year
+            "annual"  -> Just TG.Year 
+            _         -> Nothing
+          tr $ timedRecurrence g v t
       _ -> Nothing
   }
 
-ruleRecurrenceLy :: Rule
-ruleRecurrenceLy = Rule
+ruleLy :: Rule
+ruleLy = Rule
   { name = "hourly|daily|weekly|monthly|yearly"
   , pattern =
-    [regex "((bi)?(hour|dai|week|month|quarter|year|annual)ly)"
+    [regex "(hour|dai|week|month|quarter|year|annual)ly"
+    ]
+  , prod = \tokens -> case tokens of
+      (Token RegexMatch (GroupMatch (grain:_)):
+       _) -> do
+          g <- case Text.toLower grain of
+            "hour"    -> Just TG.Hour
+            "dai"     -> Just TG.Day
+            "week"    -> Just TG.Week
+            "month"   -> Just TG.Month
+            "quarter" -> Just TG.Quarter
+            "year"    -> Just TG.Year
+            "annual"  -> Just TG.Year 
+            _         -> Nothing
+          tr $ timedRecurrence g 1 1
+      _ -> Nothing
+  }
+
+ruleSemiAnnual :: Rule
+ruleSemiAnnual = Rule
+  { name = "semi-annual"
+  , pattern =
+    [regex "(bi|semi)[\\s\\-]*annual"
     ]
   , prod = \tokens -> case tokens of
       (Token RegexMatch (GroupMatch (match:_)):
+       _) -> do
+          (v, t) <- case Text.toLower match of
+            "bi"   -> Just (2, 1)
+            "semi" -> Just (1, 2)
+            _      -> Just (1, 1)
+          tr $ timedRecurrence TG.Year v t
+      _ -> Nothing
+  }
+
+ruleAnnual :: Rule
+ruleAnnual = Rule
+  { name = "annual"
+  , pattern =
+    [regex "annual"
+    ]
+  , prod = \_ -> tr $ recurrence TG.Year 1
+  }
+
+ruleEveryday :: Rule
+ruleEveryday = Rule
+  { name = "everyday"
+  , pattern =
+    [regex "everyday"
+    ]
+  , prod = \_ -> tr $ recurrence TG.Day 1
+  }
+
+ruleCompositeTimes :: Rule
+ruleCompositeTimes = Rule
+  { name = "<integer> times <recurrence>"
+  , pattern =
+    [ Predicate isNatural
+    , regex "times?"
+    , Predicate isBasicRecurrence
+    ]
+  , prod = \tokens -> case tokens of
+      (Token Numeral NumeralData{TNumeral.value}:_:Token Recurrence r:_) -> 
+        tr $ mkComposite r (floor value)
+      _ -> Nothing
+  }
+
+ruleCompositeOnce :: Rule
+ruleCompositeOnce = Rule
+  { name = "once|twice|thrice <recurrence>"
+  , pattern =
+    [ regex "(once|twice|thrice)"
+    , Predicate isBasicRecurrence
+    ]
+  , prod = \tokens -> case tokens of
+      (Token RegexMatch (GroupMatch (match:_)):Token Recurrence r:_) ->
```

Comment:
Not sure if I 'm understanding it correctly:
This is a pattern matching with `Token RegexMatch (GroupMatch (match:_))`, `Token Recurrence r` and any remaining trailing string. And the first part of pattern matching  `Token RegexMatch (GroupMatch (match:_))` should be matched with "(once|twice|thrice)" here?

Reply:
Yep, that's right